src/genetic_algorithm/algorithms/species_based_algorithm.rs

```diff
@@ -0,0 +1,176 @@
+        let mut offspring_population: Vec<T> = Vec::new();
+
+        // Create reference to elements ordered by fitness
+        // Will be used for breeding when a species has not improved for a while
+        let population = self.population.clone();
+
+        let mut population: Vec<&T> = population.values().flatten().collect();
+        population.sort_by(|a, b| b.get_fitness().partial_cmp(&a.get_fitness()).unwrap());
+
+        for (species_size, species) in self.population.iter_mut() {
+            // If the species has not improved for a while, breed it with the best individual
+            // In the first run there is no best_by_species
+            if !self.previous_best_by_species.is_empty()
+                && self.previous_best_by_species.get(species_size).unwrap().1
+                    > MAX_UNCHANGED_GENERATIONS
+            {
+                for (first_parent, second_parent) in species
+                    .iter()
+                    .zip(population.iter())
+                    .take(species.len() / 2)
+                {
+                    let (mut first_child, mut second_child) = first_parent.breed(&second_parent);
+
+                    first_child = first_child.mutate(1.0);
+                    second_child = second_child.mutate(1.0);
+                    first_child.correct();
+                    second_child.correct();
+                    offspring_population.push(first_child);
+                    offspring_population.push(second_child);
+                }
+            } else {
+                // Breed the species
+                for parents in species.chunks(2) {
+                    if parents.len() < 2 {
+                        continue;
+                    }
+
+                    let first_parent = &parents[0];
+                    let second_parent = &parents[1];
+
+                    let (mut first_child, mut second_child) = first_parent.breed(&second_parent);
+
+                    first_child = first_child.mutate(1.0);
+                    second_child = second_child.mutate(1.0);
+                    first_child.correct();
+                    second_child.correct();
+                    offspring_population.push(first_child);
+                    offspring_population.push(second_child);
+                }
+            }
+
+            if species.len() % 2 != 0 {
+                let first_parent = &species.last().unwrap();
+                let second_parent = population[species.len() - 1];
+                let (mut first_child, mut second_child) = first_parent.breed(&second_parent);
+
+                first_child = first_child.mutate(1.0);
+                second_child = second_child.mutate(1.0);
+                first_child.correct();
+                second_child.correct();
+                offspring_population.push(first_child);
+                offspring_population.push(second_child);
+            }
+        }
+
+        offspring_population.truncate(self.population.values().flatten().count());
```

Comment:
This is the number of total creatures there are, regardless of species? (Because of the `flatten`. If that's the case, would `population_size` work?

Reply:
Yes, but we don't pass population size here. So we need to use population's count instead